### PR TITLE
Extension enhancements

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/area cost
/area performance
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR contains several usability, performance and cost improvements for the `/extensions` library, mainly used by Gardener extensions
- List all controller names that can be disabled via `--disable-controllers` in the cmd help.
- Only inject shoot client into health check if requested which mainly saves API discovery calls every check period.
- Patch conditions instead of updating them to save network traffic, especially if the extension holds `.status.state`.
- Patch results of health checks in bulk instead of updating every condition type individually.

**Special notes for your reviewer**:
/cc @danielfoehrKn 

**Release note**:

```improvement operator
Several improvements for extension health checks have been implemented which increase the runtime performance and reduce network traffic.
```
```improvement developer
Several improvements for extension health checks (lib `github.com/gardener/gardener/extensions`) have been implemented which increase the runtime performance and reduce network traffic. Please update your `github.com/gardener/gardener` to the latest version to profit from these changes.
```

### CODE REVIEW

File: extensions/pkg/controller/utils.go [hunk 1]
```diff
@@ -206,6 +206,51 @@ func GetSecretByReference(ctx context.Context, c client.Client, ref *corev1.Secr
...
+// It retries the patch with an exponential backoff.
+func TryPatch(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, transform func() error) error {
+	return tryPatch(ctx, backoff, c, obj, c.Patch, transform)
+}
+
+// TryPatchStatus tries to apply the given transformation function onto the given object, and to patch its
+// status afterwards with optimistic locking. It retries the status patch with an exponential backoff.
+func TryPatchStatus(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, transform func() error) error {
+	return tryPatch(ctx, backoff, c, obj, c.Status().Patch, transform)
+}
```
Comment (rfranzke) [inline extensions/pkg/controller/utils.go]: Does it make sense to add some unit tests for those?

**Review by rfranzke — COMMENTED** [review]
Generally /lgtm

File: extensions/pkg/controller/utils.go (hunk 1 above)
Comment (timuthy) [reply extensions/pkg/controller/utils.go]: Yep, I'll add some.

File: extensions/pkg/controller/healthcheck/general/managed_resource.go [hunk 2]
```diff
@@ -49,11 +48,6 @@ func (healthChecker *ManagedResourceHealthChecker) InjectSeedClient(seedClient c
 	healthChecker.seedClient = seedClient
 }
 
-// InjectShootClient injects the shoot client
-func (healthChecker *ManagedResourceHealthChecker) InjectShootClient(shootClient client.Client) {
```
Comment (danielfoehrKn) [inline extensions/pkg/controller/healthcheck/general/managed_resource.go:53]: nice, so for managed resource health checks, a shoot client is not created any more?

**Review by danielfoehrKn — DISMISSED** [review]
Looks good to me. Thanks!

File: extensions/pkg/controller/healthcheck/general/managed_resource.go (hunk 2 above)
Comment (timuthy) [reply extensions/pkg/controller/healthcheck/general/managed_resource.go:53]: Yes, a health checker now needs to have a `InjectShootClient` function to get the client.

File: extensions/pkg/controller/utils.go (hunk 1 above)
Comment (timuthy) [reply extensions/pkg/controller/utils.go]: I added unit test, similar to `tryUpdate` and at the same time removed the code from the `/extensions` dir. Now only a func variable points to the implementation in `kutil`. I hope that's ok.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**